File: cloud-functions/lib/summary.py
import random

TEMPLATES_KELEBIHAN = [
    "Visi {nama} skor {skor}% — paling nyamip dengan harapan Anda.",
    "Dari segi {kategori}, {nama} unggul banget: skor {skor}%.",
    "Skor {kategori} {nama} ({skor}%) jadi nilai jual utama.",
    "{nama} paling menonjol di {kategori}, skor {skor}%.",
    "{kategori.title()} {nama} (skor {skor}%) patut diacungi jempol.",
    "Kelebihan utama {nama} ada di {kategori} dengan skor {skor}%.",
    "Nilai {kategori} {nama} ({skor}%) termasuk yang terbaik.",
    "{nama} cocok banget dari sisi {kategori}: skor {skor}%.",
    "{kategori.title()} jadi senjata utama {nama} — skor {skor}%.",
    "Luar biasa, skor {kategori} {nama} tembus {skor}%.",
    "Di antara semua calon, {nama} paling unggul di {kategori} ({skor}%).",
    "Skor {kategori} {nama} ({skor}%) bikin dia beda dari yang lain.",
    "{nama} punya nilai plus di {kategori}, skor {skor}%.",
    "Kalau soal {kategori}, {nama} juaranya: skor {skor}%.",
    "Gak bisa dipungkiri, {kategori} {nama} jadi andalan ({skor}%).",
    "Dari {kategori}, {nama} dapat skor {skor}% — mantap.",
    "Satu hal yang bikin {nama} unggul: {kategori} skor {skor}%.",
    "{nama} bersinar di {kategori}, skor {skor}%.",
    "Skor {kategori} {nama} ({skor}%) bikin dia layak dipertimbangkan.",
    "Keunggulan {nama} terlihat jelas di {kategori}: skor {skor}%.",
]

TEMPLATES_KELEMAHAN = [
    "Catatan: skor {kategori} {nama} masih rendah ({skor}%).",
    "Kelemahan {nama} ada di {kategori}, skor cuma {skor}%.",
    "Sayangnya, {kategori} {nama} kurang memenuhi ({skor}%).",
    "Skor {kategori} {nama} ({skor}%) masih perlu diperbaiki.",
    "Perlu dicatat, {kategori} bukan point plus {nama} ({skor}%).",
    "{nama} kurang greget di {kategori}, skor {skor}%.",
    "Dari sisi {kategori}, {nama} dapat skor {skor}% — masih kurang.",
    "Evaluasi: {kategori} {nama} cuma {skor}%, di bawah rata-rata.",
    "{kategori.title()} jadi titik lemah {nama} dengan skor {skor}%.",
    "Yang perlu diwaspadai: skor {kategori} {nama} baru {skor}%.",
    "Skor {kategori} {nama} ({skor}%) belum sesuai harapan.",
    "{nama} masih perlu kerja keras di {kategori} ({skor}%).",
    "Untuk {kategori}, {nama} masih kurang skor ({skor}%).",
    "Hasil {kategori} {nama} belum maksimal: skor {skor}%.",
    "Skor {kategori} {nama} ({skor}%) masih bisa ditingkatkan.",
    "Catatan kecil: {kategori} {nama} skor {skor}%, perlu perhatian.",
    "Dari 4 aspek, {kategori} jadi nilai terendah {nama} ({skor}%).",
    "Yang bikin {nama} turun skor: {kategori} cuma {skor}%.",
    "Kekurangan {nama} paling terasa di {kategori} ({skor}%).",
    "Skor {kategori} {nama} ({skor}%) masih di bawah ekspektasi.",
]

TEMPLATES_PERINGKAT = [
    "Peringkat 1 unggul tipis {selisih}% dari peringkat 2.",
    "Dominasi penuh! Peringkat 1 unggul {selisih}% dari runner-up.",
    "Selisih {selisih}% — peringkat 1 dan 2 bersaing ketat.",
    "Peringkat 1 unggul cukup jauh, selisih {selisih}%.",
    "Kompetisi ketat: peringkat 1 hanya unggul {selisih}%.",
    "Peringkat 1 memimpin dengan keunggulan {selisih}%.",
    "Jarak peringkat 1 dan 2: {selisih}%.",
    "Peringkat 1 nyaman di puncak dengan selisih {selisih}%.",
    "Selisih tipis {selisih}%, peringkat 2 masih bisa mengejar.",
    "Peringkat 1 kokoh dengan selisih {selisih}% dari peringkat 2.",
    "Unggul {selisih}% — peringkat 1 layak jadi andalan.",
    "Peringkat 1 mempertahankan posisi dengan selisih {selisih}%.",
    "Hanya berselisih {selisih}%, hasil bisa berbeda lain waktu.",
    "Peringkat 1 memimpin {selisih}%, tapi peringkat 2 patut diperhitungkan.",
    "{selisih}% jadi jarak antara peringkat 1 dan 2.",
    "Peringkat 1 lebih cocok {selisih}% dibanding peringkat 2.",
    "Keunggulan {selisih}% bikin peringkat 1 tak tergoyahkan.",
    "Meski unggul {selisih}%, peringkat 1 tetap perlu diwaspadai.",
    "Peringkat 2 tertinggal {selisih}% dari peringkat 1.",
    "Pertarungan ketat — cuma beda {selisih}% antara peringkat 1 dan 2.",
]

TEMPLATES_TIP = [
    "Coba atur ulang range umur buat hasil yang beda.",
    "Mau coba lagi? Ubah prioritas misi pilihan Anda.",
    "Tips: naikkan pendidikan minimal lihat calon baru muncul.",
    "Coba isi visi yang lebih spesifik biar skor makin akurat.",
    "Ubah-ubah pilihan misi, lihat perubahan ranking-nya.",
    "Cek hasil lagi setelah ganti preferensi umur.",
    "Mau coba skenario lain? Ubah visi ideal Anda.",
    "Eksperimen dengan beda kombinasi misi dan lihat hasilnya.",
    "Pilih lebih banyak misi atau kurangi, ranking bisa berubah.",
    "Atur ulang prioritas — hasil bisa beda tiap percobaan.",
    "Coba bandingkan peringkat 1 dan 2 di halaman compare.",
    "Jangan puas sama satu hasil, coba kombinasi lain.",
    "Semakin detail visi Anda, semakin akurat hasilnya.",
    "Pendidikan minimal bisa bikin perbedaan besar — coba ubah.",
    "Coba rentang umur yang lebih sempit buat hasil presisi.",
    "Mainkan bobot preferensi Anda dan lihat perubahannya.",
    "Gunakan fitur compare buat lihat beda tiap calon.",
    "Ubah strategi: fokus ke 1-2 misi prioritas saja.",
    "Bandingkan hasil dengan pilih calon idola Anda langsung.",
    "Coba refresh halaman dan isi ulang — hasil mungkin beda.",
]

KATEGORI_MAP = {
    "skor_visi": "visi",
    "skor_misi": "visi misi",
    "skor_pendidikan": "pendidikan",
    "skor_umur": "umur",
}


def _kategori_label(key: str) -> str:
    return KATEGORI_MAP.get(key, key.replace("skor_", ""))


def buat_summary(results: list) -> list:
    summaries = []
    top_score = results[0].skor_total if results else 0

    for i, r in enumerate(results):
        parts = []

        # Sort scores to find strength & weakness
        scores = [
            ("skor_visi", r.skor_visi),
            ("skor_misi", r.skor_misi),
            ("skor_pendidikan", r.skor_pendidikan),
            ("skor_umur", r.skor_umur),
        ]
        scores_sorted = sorted(scores, key=lambda x: x[1], reverse=True)
        tertinggi = scores_sorted[0]
        terendah = scores_sorted[-1]

        # Strength
        kat_baik = _kategori_label(tertinggi[0])
        t = random.choice(TEMPLATES_KELEBIHAN).format(
            nama=r.nama.split(",")[0].split()[0] if "," in r.nama else r.nama.split()[0],
            skor=int(tertinggi[1]),
            kategori=kat_baik,
        )
        parts.append(t)

        # Weakness (if not perfect)
        if terendah[1] < 100:
            kat_buruk = _kategori_label(terendah[0])
            t = random.choice(TEMPLATES_KELEMAHAN).format(
                nama=r.nama.split(",")[0].split()[0] if "," in r.nama else r.nama.split()[0],
                skor=int(terendah[1]),
                kategori=kat_buruk,
            )
            parts.append(t)

        # Rank 1 comparison (only for rank 1)
        if i == 0 and len(results) > 1:
            selisih = r.skor_total - results[1].skor_total
            t = random.choice(TEMPLATES_PERINGKAT).format(selisih=f"{selisih:.0f}")
            parts.append(t)

        # Tip (only for rank 1, every 3rd request-ish)
        if i == 0:
            t = random.choice(TEMPLATES_TIP)
            parts.append(t)

        summaries.append(" ".join(parts))

    return summaries

File: src/lib/summary.ts
function pilih<T>(arr: T[]): T {
  return arr[Math.floor(Math.random() * arr.length)]
}

const PREAMBLES = [
  "Dari hasil analisis,", "Berdasarkan preferensi Anda,",
  "Kalau dilihat dari data yang ada,", "Menurut perhitungan kecocokan,",
  "Dari semua calon yang ada,", "Secara keseluruhan,", "Bila dilihat dari kriterianya,",
]

export function buatSummary(params: {
  nama: string; nomor_urut: number
  skor_visi: number; skor_misi: number
  skor_pendidikan: number; skor_umur: number
  skor_total: number
  visi_calon: string; misi_calon: string[]
  pendidikan_calon: string; umur_calon: number
  misi_user_set: Set<string>
  pendidikan_min: string; umur_min: number; umur_max: number
  ranking: number; total_calon: number; selisih_atas: number | null
}): string {
  const {
    nama, skor_visi, skor_misi, skor_pendidikan, skor_umur, skor_total,
    misi_calon, misi_user_set, pendidikan_calon, umur_calon,
    pendidikan_min, umur_min, umur_max, ranking, total_calon, selisih_atas,
  } = params

  const nd = nama.split(",")[0]?.split(" ")[0] ?? nama.split(" ")[0]
  const parts: string[] = [pilih(PREAMBLES)]

  if (skor_visi >= 70) {
    parts.push(pilih([
      `visi ${nd} (${skor_visi.toFixed(0)}%) cukup sejalan dengan gambaran Anda.`,
      `visi calon ini mirip dengan harapan Anda - skor ${skor_visi.toFixed(0)}%.`,
      `kesamaan visi cukup tinggi (${skor_visi.toFixed(0)}%), ${nd} punya arah yang sama dengan Anda.`,
    ]))
  } else if (skor_visi >= 35) {
    parts.push(pilih([
      `visi ${nd} cukup dekat (${skor_visi.toFixed(0)}%), walau ada beberapa perbedaan.`,
      `skor visi ${skor_visi.toFixed(0)}% - ada titik temu, tapi belum sepenuhnya sejalan.`,
      `visi ${nd} punya kesamaan secukupnya dengan Anda (${skor_visi.toFixed(0)}%).`,
    ]))
  } else {
    parts.push(pilih([
      `visi ${nd} masih cukup berbeda dari harapan Anda (${skor_visi.toFixed(0)}%).`,
      `skor visinya ${skor_visi.toFixed(0)}% - arah pemikirannya belum terlalu cocok.`,
      `visi yang Anda tulis belum banyak nyambung dengan visi ${nd} (${skor_visi.toFixed(0)}%).`,
    ]))
  }

  const sama = misi_calon.filter(m => misi_user_set.has(m))
  if (sama.length > 0) {
    const acak = sama.sort(() => Math.random() - 0.5).slice(0, 2)
    if (acak.length === 1) {
      parts.push(pilih([
        `Misi '${acak[0].slice(0, 50)}...' juga jadi prioritas Anda.`,
        `Salah satu misi andalan Anda - '${acak[0].slice(0, 50)}...' - juga diusung ${nd}.`,
        `Ada kesamaan di misi: '${acak[0].slice(0, 50)}...'.`,
      ]))
    } else {
      parts.push(pilih([
        `Dua misi juga Anda prioritaskan: '${acak[0].slice(0, 40)}...' dan '${acak[1].slice(0, 40)}...'.`,
        `Beberapa misi cocok, misalnya '${acak[0].slice(0, 40)}...' dan '${acak[1].slice(0, 40)}...'.`,
        `Misi seperti '${acak[0].slice(0, 40)}...' dan '${acak[1].slice(0, 40)}...' sejalan dengan pilihan Anda.`,
      ]))
    }
  } else if (skor_misi >= 60) {
    parts.push(pilih([
      `Misi-misi ${nd} secara umum sejalan dengan prioritas Anda (${skor_misi.toFixed(0)}%).`,
      `Secara umum arah misi ${nd} cocok dengan Anda (${skor_misi.toFixed(0)}%).`,
    ]))
  } else if (skor_misi < 30) {
    parts.push(pilih([
      `Misi ${nd} belum banyak yang sesuai dengan prioritas Anda (${skor_misi.toFixed(0)}%).`,
      `Skor misi ${skor_misi.toFixed(0)}% - arah programnya masih berbeda dengan harapan Anda.`,
    ]))
  }

  if (skor_pendidikan >= 100) {
    parts.push(pilih([
      `Pendidikan ${nd} (${pendidikan_calon}) sudah di atas batas minimal yang Anda tentukan.`,
      `Dari sisi pendidikan, ${pendidikan_calon} - memenuhi syarat yang Anda cari.`,
      `Latar belakang pendidikan ${pendidikan_calon} sudah sesuai standar Anda.`,
    ]))
  } else if (skor_pendidikan >= 50) {
    parts.push(pilih([
      `Pendidikan ${pendidikan_calon} - lumayan mendekati batas minimal (${pendidikan_min}).`,
      `Meski belum mencapai ${pendidikan_min}, pendidikan ${nd} (${pendidikan_calon}) masih cukup.`,
    ]))
  } else {
    parts.push(pilih([
      `Pendidikan ${pendidikan_calon} mungkin masih kurang dari yang Anda harapkan (${pendidikan_min}).`,
      `Catatan: pendidikan ${nd} (${pendidikan_calon}) di bawah minimal (${pendidikan_min}) yang Anda tetapkan.`,
    ]))
  }

  if (skor_umur >= 100) {
    parts.push(pilih([
      `Usia ${umur_calon} tahun pas dengan range yang Anda cari (${umur_min}-${umur_max}).`,
      `Usia ${umur_calon} tahun - tepat dalam rentang usia ideal Anda.`,
      `Umur ${nd} (${umur_calon} tahun) masuk range yang Anda inginkan.`,
    ]))
  } else if (skor_umur >= 50) {
    parts.push(pilih([
      `Usia ${umur_calon} tahun - cukup dekat dengan rentang (${umur_min}-${umur_max}).`,
      `Umurnya ${umur_calon} tahun, sedikit di luar preferensi usia Anda.`,
    ]))
  } else {
    parts.push(pilih([
      `Usia ${umur_calon} tahun - cukup jauh dari rentang usia yang Anda inginkan (${umur_min}-${umur_max}).`,
      `Catatan usia: ${umur_calon} tahun - agak meleset dari range ideal Anda.`,
    ]))
  }

  if (ranking === 1) {
    if (total_calon > 1 && selisih_atas != null) {
      if (selisih_atas < 5) {
        parts.push(pilih([
          `Peringkat 1, tapi selisihnya tipis banget - beda ${selisih_atas.toFixed(0)}% aja.`,
          `Nomor satu! Tapi hati-hati, selisih dengan peringkat 2 cuma ${selisih_atas.toFixed(0)}%.`,
          `Paling cocok, walau persaingannya ketat (selisih ${selisih_atas.toFixed(0)}%).`,
        ]))
      } else if (selisih_atas >= 15) {
        parts.push(pilih([
          `Peringkat 1 dengan skor ${skor_total.toFixed(0)}% - unggul jauh dari yang lain (${selisih_atas.toFixed(0)}%).`,
          `Jelas paling unggul dengan selisih ${selisih_atas.toFixed(0)}% dari peringkat 2.`,
          `Dominasi penuh - skor ${skor_total.toFixed(0)}%, beda ${selisih_atas.toFixed(0)}% dari pesaing terdekat.`,
        ]))
      } else {
        parts.push(pilih([
          `Peringkat 1 dengan skor ${skor_total.toFixed(0)}%, unggul ${selisih_atas.toFixed(0)}% dari peringkat 2.`,
          `Keluar sebagai yang paling cocok (${skor_total.toFixed(0)}%), dengan selisih ${selisih_atas.toFixed(0)}%.`,
        ]))
      }
    } else {
      parts.push(pilih([
        `Satu-satunya calon - skor kecocokan ${skor_total.toFixed(0)}%.`,
        `Hanya ada satu calon, skor kecocokan ${skor_total.toFixed(0)}%.`,
      ]))
    }
  } else if (ranking === 2 && total_calon > 1) {
    parts.push(pilih([
      `Peringkat 2 dari ${total_calon} calon dengan skor ${skor_total.toFixed(0)}%.`,
      `Berada di posisi kedua (${skor_total.toFixed(0)}%), masih patut dipertimbangkan.`,
    ]))
  } else if (ranking >= 3) {
    parts.push(pilih([
      `Peringkat ke-${ranking} dari ${total_calon} calon (${skor_total.toFixed(0)}%).`,
      `Cukup sulit bersaing di peringkat ${ranking} dengan skor ${skor_total.toFixed(0)}%.`,
    ]))
  }

  if (ranking === 1) {
    const kurang = misi_calon.filter(m => !misi_user_set.has(m))
    if (kurang.length > 0 && skor_misi < 80) {
      parts.push(pilih([
        "Coba tambah atau kurangi pilihan misi, hasil bisa beda.",
        "Mau variasi? Ubah-ubah prioritas misi Anda.",
        "Atur ulang preferensi misi untuk lihat perubahan ranking.",
      ]))
    } else {
      parts.push(pilih([
        "Gunakan fitur compare untuk lihat perbandingan langsung.",
        "Coba bandingkan dengan calon lain di fitur Bandingkan.",
        "Lihat detail calon lain dengan fitur compare.",
      ]))
    }
  }

  const s = parts.join(" ")
  return s.charAt(0).toUpperCase() + s.slice(1)
}
